lib/elixir/lib/calendar/naive_datetime.ex

```diff
@@ -270,7 +270,7 @@ defmodule NaiveDateTime do
   """
   @doc since: "1.4.0"
   @spec add(t, integer, System.time_unit()) :: t
-  def add(%NaiveDateTime{} = naive_datetime, integer, unit \\ :second)
+  def add(naive_datetime, integer, unit \\ :second)
```

Comment:
Why this change? I believe the checking for NaiveDateTime is intentional due to the inherent loss of precision.

Reply:
The same reason you can pass a DateTime or NaiveDateTime to `Date.leap_year?/1` or `Date.to_iso8601` etc. We started to change the functions to accept e.g. `Calendar.date` instead of only `Date`. But unfortunately some functions were not checked/changed. For consistency we can change the rest.

In this case I see that in the specs `t` should be changed to `Calendar.naive_datetime`.

Reply:
We agreed though, and I think it was your suggestion, to not allow them in `add/2` because of the inherent loss of precision that happens when you pass a `DateTime` to `NaiveDateTime.add`. For example offsets are not considered. So mostly we allow `Calendar.naive_datetime` on read-only operations and the struct on write ones.

Still, I think if we want to remove this restriction, we should probably do it consistently and in another pull request, as this one is already getting quite large. :)